### PR TITLE
Just another fix for DDLWorker

### PR DESCRIPTION
**Pull request by tavplubix**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Detailed description / Documentation draft:
DDLWorker might skip or execute twice some queue entries.
Release branches are not affected (yet).
Introduced in #16193 and #21079
cc: @azat

### CODE REVIEW

File: src/Interpreters/DDLWorker.cpp [hunk 1]
```diff
@@ -311,51 +312,117 @@ void DDLWorker::scheduleTasks(bool reinitialized)
...
+                else
                     processTask(*task, zookeeper);
-                }
+                ++task_it;
+            }
+            else
+            {
+                /// We didn't even executed a query, so let's just remove it.
+                /// We will try to read the task again and execute it from the beginning.
+                if (!first_failed_task_name || task->entry_name < *first_failed_task_name)
```
Comment (azat) [inline src/Interpreters/DDLWorker.cpp:357]: Maybe `DDLTaskBase::entry_name` worth a comment that it is safe to compare strings, since strings are padded?

**Review by azat — APPROVED** [review]
BTW is there a chance to cover this with an integration test?

**Comment by tavplubix** [comment]
> BTW is there a chance to cover this with an integration test?

I'm not sure, integration tests are complicated, too hard to write and not reliable enough. All tests passed in #16193 and #21079, but then we noticed random failures of `test_distributed_ddl` in master.

File: src/Interpreters/DDLWorker.cpp [hunk 2]
```diff
@@ -311,51 +312,120 @@ void DDLWorker::scheduleTasks(bool reinitialized)
...
-                bool task_still_exists = zookeeper->exists(task->entry_path);
+                /// Connection was lost on attempt to write status. Will retry.
                 bool status_written = zookeeper->exists(task->getFinishedNodePath());
-                if (!status_written && task_still_exists)
-                {
+                /// You might think that the following condition is redundant, because status_written implies completely_processed.
+                /// But it's wrong. It's possible that (!task->completely_processed && status_written)
+                /// if ZooKeeper successfully received and processed our request
+                /// but we lost connection while waiting for the response.
+                /// Yeah, distributed systems is a zoo.
```
Comment (alesapin) [inline src/Interpreters/DDLWorker.cpp:346]: :D

**Pull request merged**